bugsnag-android-core/src/main/java/com/bugsnag/android/internal/DocumentPath.kt

```diff
@@ -0,0 +1,209 @@
+                        }
+                        directives.add(makeDirectiveFromPathComponent(pathComponent))
+                        strBegin = i + 1
+                        requiresEscaping = false
+                    }
+                    else -> {
+                    }
+                }
+                i++
+            }
+            if (strBegin < path.length) {
+                val pathComponent: String = if (requiresEscaping) {
+                    unescape(path, strBegin, path.length)
+                } else {
+                    path.substring(strBegin)
+                }
+                directives.add(makeDirectiveFromPathComponent(pathComponent))
+            }
+            if (isDotLast(path)) {
+                directives.add(DocumentPathDirective.ListInsertDirective())
+            }
+            return directives
+        }
+
+        private fun unescape(str: String, startOffset: Int, endOffset: Int): String {
+            val buff = StringBuilder(endOffset - startOffset)
+            var i = startOffset
+            while (i < endOffset) {
+                val ch = str[i]
+                if (ch == '\\') {
+                    i++
+                    require(i < endOffset) { "Path cannot end on an escape char '\\'" }
+                    buff.append(str[i])
+                } else {
+                    buff.append(ch)
+                }
+                i++
+            }
+            return buff.toString()
+        }
+
+        private fun isDotLast(path: String): Boolean {
+            if (path[path.length - 1] != '.') {
+                return false
+            }
+            if (path.length == 1) {
+                return true
+            }
+            var isUnescapedDot = true
+            for (i in path.length - 2 downTo 0) {
+                val ch = path[i]
+                isUnescapedDot = if (ch == '\\') {
+                    !isUnescapedDot
+                } else {
+                    break
+                }
+            }
+            return isUnescapedDot
+        }
+
+        private fun makeDirectiveFromPathComponent(pathComponent: String?): DocumentPathDirective {
+            if (isInteger(pathComponent)) {
+                val index = pathComponent!!.toInt()
+                return if (index == -1) {
+                    DocumentPathDirective.ListLastIndexDirective()
+                } else {
+                    DocumentPathDirective.ListIndexDirective(index)
+                }
+            }
+            return DocumentPathDirective.MapKeyDirective(pathComponent!!)
+        }
+
+        private fun isInteger(str: String?): Boolean {
+            if (str!!.isEmpty()) {
+                return false
+            }
+            var startIndex = 0
+            if (str[0] == '-') {
+                startIndex = 1
+            }
+            for (i in startIndex until str.length) {
+                val ch = str[i].toInt()
+                if (ch < '0'.toInt() || ch > '9'.toInt()) {
```

Comment:
Is it worth caching the character values for `0` and `9` to avoid the re-evaluations here for each character in `str` (in the pathological case)?

Reply:
Probably not - `toInt` is a compiler intrinsic in these cases:

```
    LINENUMBER 202 L15
    ILOAD 5
    BIPUSH 48
    IF_ICMPLT L16
    ILOAD 5
    BIPUSH 57
    IF_ICMPLE L17
```

Reply:
aha! thank you

Reply:
I would have thought that the compiler is smart enough to do hoist these values out automatically? I could defensively guard against it just in case.

Reply:
I tend to skew skeptical of new-ish compilers, but having taken a look, I'm no longer concerned.